Subnet range selector, interface fixes

0xCA opened the pull request:
# Subnet range selector
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/11984323/1b5d4fdd-2ead-4cd6-a63c-d4165f4a7273)

> 10.0.0.1/16 server with a very few clients, correctly suggesting an IP from 10.0.4.0/22 for a new client

## Reason
This feature is intended for subdivision of large wireguard servers, such as /16 subnets.
I wanted to allocate smaller blocks of the server address space for specific purposes and easier firewalling.
Current IP suggestion is unaware of such blocks, which requires entering each IP manually.
So I made it aware.
## Details
- Subnet ranges are set in an **optional** variable `SUBNET_RANGES`
- Format: `SR Name| 10.0.1.0/24; SR2| 10.0.2.0/24, 10.0.3.0/24, 2001:0db8::/32`
- Each CIDR must be inside one of the server interfaces
- Other than that, you can go for a really small CIDRs, no problem
- Any invalid CIDRs or entire SRs are discarded during the server load
- The list is only stored in memory, so you can change it by changing the env variable and restarting wireguard-ui
- Multiple server interfaces or multiple client CIDRs are supported, but not auto-suggested
- **No part of this feature is saved to the database**
### IP suggestion
- IP is suggested from a chosen SR
- If no free IPs are left, an error is shown
- Default SR is Any, which is "any server interface", same as no SR at all
- **For existing clients**, you can select a different SR, a new IP is automatically suggested
- If you select the same SR, a new IP is suggested
- Fix for multiple interfaces: it only fails if no interface has free IPs
### Subnet range filter
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/11984323/6b9a8ef4-3b78-4561-8a2d-cb2ddc708dd3)
- Allows to show clients from a selected subnet range for easier managing
- Pure client side, no requests
- A SR is detected server-side when wg client info is requested
- Only the first match is used
- There is an in-memory lookup cache on the server, that reduces subsequent cidr-contains-ip evaluations

## Extra
![image](https://github.com/ngoduykhanh/wireguard-ui/assets/11984323/3fc95513-e1a6-4771-87ee-0e5ba16cf63a)
I also Implemented some interface/UX fixes
- Apply config button is properly shown or hidden when needed
- Toast notifications are not obstructing the buttons anymore

Commit:
Parse and validate subnet ranges

Commit:
Full implementation of subnet ranges for a new client, fixes

fix: free ip search was stopping after no free ip on the first interface
fix: toast obstructing the buttons
fix: stuck apply config button

Commit:
Added clients page filter by subnet range

Commit:
Further Apply config visibility fixes. Fixed tag input being too small.

Commit:
Added semiauto subnet range changer for existing clients

If the same subnet range is selected, a different IP is suggested

Commit:
Added SUBNET_RANGES to readme

Commit:
A bit more usable makefile

nebulosa2007 commented:
You done great work, but you want implement several options at one time and it's not convinient for development..
Could you split you changes on three branches:
1. ip-range-selector
2. interface/UX fixes
3. adding a new makefile (if it new file really needed for options above)

It will increase the chances of appearing in the main branch

Commit:
Revert tag input being too small fix

systemcrash commented:
@nebulosa2007 is right here. 
-Remove the makefile additions. They're not useful in this repo. (maybe for you to compile locally)
-Separate the UX fixes from the go code fixes for the SR stuff.

systemcrash commented on Makefile:
```diff
@@ -0,0 +1,11 @@
+build-ui:
```
Drop the makefile

systemcrash commented on README.md:
```diff
@@ -37,10 +37,11 @@ docker-compose up
 
 | Variable                    | Description                                                                                                                                                  | Default                            |
-|-----------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------|
+| --------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ | ---------------------------------- |
```
drop these formatting fixes

systemcrash commented on README.md:
```diff
@@ -37,10 +37,11 @@ docker-compose up
-| `BIND_ADDRESS`              | The addresses that can access to the web interface and the port                                                                                              | 0.0.0.0:80                        |
+| `BIND_ADDRESS`              | The addresses that can access to the web interface and the port                                                                                              | 0.0.0.0:80                         |
 | `SESSION_SECRET`            | The secret key used to encrypt the session cookies. Set this to a random value                                                                               | N/A                                |
+| `SUBNET_RANGES`             | The list of address subdivision ranges. Format: `SR Name: 10.0.1.0/24; SR2: 10.0.2.0/24, 10.0.3.0/24` Each CIDR must be inside one of the server interfaces. | N/A                                |
```
Just leave the important stuff. 

Test also to drop that example string into the docker-compose.yaml file. Error. 

Test it on https://www.yamllint.com/

```yaml
...
    environment:
      - SUBNET_RANGES=SR Name:10.0.1.0/24; SR2:10.0.2.0/24,10.0.3.0/24
...
```

Clean it up so that it works.

systemcrash replied:
Because that's where it will also get used. Why make it harder for users to
get started? Friction.

systemcrash commented on handler/routes.go:
```diff
@@ -358,17 +362,17 @@ func GetClient(db store.IStore) echo.HandlerFunc {
-			Enabled:       true,
-			IncludeDNS:    true,
-			IncludeMTU:    true,
+			Enabled:    true,
```
Drop changes like this. They can be fixed later. They only add noise to this PR.

systemcrash replied:
In a separate commit, in a separate PR.

It's default gofmt. It should be this way anyway.

systemcrash replied:
I've made PRs which waited for years, even tho the repo is active. But reviewers don't have unlimited time to pick apart your PR. The first question they ask themselves is: how does this formatting affect this new functionality? Go is a bit weird where indentation is pedantic.

Don't conflate fixing someone else's formatting omission with your PR.


Once the PR gets accepted, do a new one with go fmt results if the repo owner doesn't do this themselves.

systemcrash commented on handler/routes.go:
```diff
@@ -486,9 +490,9 @@ func EmailClient(db store.IStore, mailer emailer.Emailer, emailSubject, emailCon
-			Enabled:       true,
-			IncludeDNS:    true,
-			IncludeMTU:    true,
+			Enabled:    true,
```
Drop...

systemcrash commented on templates/base.html:
```diff
@@ -364,7 +372,37 @@ <h1>{{template "page_title" .}}</h1>
 
-                $.ajax({
+            addGlobalStyle(`
+.toast-top-right-fix {
```
Undo this indent adjustment. Keep the beef of this PR separate from formatting changes.

systemcrash commented:
Because that's where it will also get used. Why make it harder for users to
get started? Friction.


> It works just fine in .env
> Why placing it in the compose file?
> Spaces are there for readability. If they also prevent hard-configs, the
> better.
>

systemcrash commented:
In a separate commit, in a separate PR.

 It's default gofmt. It should be this way anyway.
>

systemcrash commented:
Nobody likes to go down rabbit holes to figure stuff out. Present the
information.

What should one look for? The html file is big.

>
> Do you mean it should be clarified? It's obvious if you follow it. This
> html section is completely overridden in that js function.
>

systemcrash commented:
We would not ask for it if it wasn't worth it.

It makes reviewing this and any PR painful when we need to separate
formatting from the noise of what's actually important.

Rule: Put them in separate commits, in a separate PR.

Then it's easy for anyone to understand what the PR is supposed to do.

I like the core of this PR, otherwise.


> Is it really worth it? I do understand that it might prevent a minor
> conflict here and there. But it significantly degrades the DX. OR requires
> a manual "second pass" with all the formatters turned off (which is not
> particularly convenient to achieve, let alone the process of manually
> reverting every fmt change).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ngoduykhanh/wireguard-ui/pull/481#issuecomment-1837551792>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6DULTD6D2VORMI4VMAWLYHS4APAVCNFSM6AAAAAA66W2LKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMZXGU2TCNZZGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

systemcrash commented:
Also, if for some reason anyone needs to revert this PR, they don't need to
deal with the fallout of formatting unexpectedly removed which affects
commits on top of it also.

So dedicate a PR to running go fmt on the code base. 👌


> Is it really worth it? I do understand that it might prevent a minor
> conflict here and there. But it significantly degrades the DX. OR requires
> a manual "second pass" with all the formatters turned off (which is not
> particularly convenient to achieve, let alone the process of manually
> reverting every fmt change).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ngoduykhanh/wireguard-ui/pull/481#issuecomment-1837551792>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6DULTD6D2VORMI4VMAWLYHS4APAVCNFSM6AAAAAA66W2LKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMZXGU2TCNZZGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

systemcrash commented:
Sorry, on email.

Could you please keep your answers in their threads?
>

systemcrash commented on templates/base.html:
```diff
@@ -58,11 +58,13 @@
                 </div>
                 <div class="form-group form-group-sm">
                     <select name="status-selector" id="status-selector" class="custom-select form-control-navbar" style="margin-left: 0.5em; height: 90%; font-size: 14px;">
+                        <!-- SEE updateSearchList() in clients.html BEFORE EDITING -->
```
Nobody likes to go down rabbit holes to figure stuff out. Present the
information.

What should one look for? The html file is big.

Commit:
Fixed issues mentioned in the @systemcrash review

- Reverted formatting fixes where possible
- Clarified the note about updateSearchList()
- Fixed a typo
- Made SUBNET_RANGES example yaml-friendly
- Removed Makefile

ngoduykhanh reviewed (commented):
Thank you @0xCA for the PR and everybody here for your contribution to making the PR better. Overall, the changes look good to me. I've just left one comment. I'm happy to merge it if @0xCA can follow up here and resolve the conflicts.

Commit:
Fix duplicated IP suggestions

ngoduykhanh commented on templates/base.html:
```diff
@@ -209,6 +211,12 @@ <h4 class="modal-title">New Wireguard Client</h4>
                                 <label for="client_email" class="control-label">Email</label>
                                 <input type="text" class="form-control" id="client_email" name="client_email">
                             </div>
+                            <div class="form-group">
+                                <label for="subnet_ranges" class="control-label">Subnet range</label>
+                                <select id="subnet_ranges" class="select2"
+                                    data-placeholder="Select a subnet range" style="width: 100%;">
+                                </select>
+                            </div>
```
IDK, it was my personal preference. No worries, I can keep it like this and refactor it later.
Leaving it in the "Additional configuration" doesn't seem nice to me as I think it should be part of the IP allocation section.

ngoduykhanh replied:
Endpoint is a standard configuration supported by WireGuard, isn't it?

ngoduykhanh commented:
@0xCA I replied to your comment and merged this PR. Thanks again!
